### PR TITLE
MOSIP-26717 There is no mechanism to re-trigger failed notifications or notifications that are not sent

### PR DESCRIPTION
**Pull request by Khuddusshariff0022**

### CODE REVIEW

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/constant/APIAuthorityList.java [hunk 1]
```diff
@@ -30,6 +30,8 @@ public enum APIAuthorityList {
 
 	REQUESTHANDLER(new String[] { "REGISTRATION_ADMIN", "REGISTRATION_PROCESSOR" }),
 
+	USERNOTIFICATION(new String[] { "REGISTRATION_PROCESSOR" }),
```
Comment (vishwa-vyom) [inline registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/constant/APIAuthorityList.java]: Should we not include the REGISTRATION_ADMIN here ?

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/exception/util/PlatformErrorMessages.java [hunk 2]
```diff
@@ -1403,6 +1403,19 @@ public enum PlatformErrorMessages {
...
+			"Missing Input Parameter - %s"),
+
+	/** The invalid input parameter. */
+	RPR_UNA_INVALID_INPUT_PARAMETER(PlatformConstants.RPR_USER_NOTIFICATION_API + "002",
+			"Invalid Input Parameter - %s"),
+	RPR_UNA_RID_NOT_FOUND(PlatformConstants.RPR_USER_NOTIFICATION_API + "003",
+			"Rid not found in registration table"),
+	RPR_UNA_UNKNOWN_EXCEPTION(PlatformConstants.RPR_USER_NOTIFICATION_API + "004",
+			"Unknown exception occured"),
+	RPR_UNA_REGTYPE_NOT_MATCHING(PlatformConstants.RPR_USER_NOTIFICATION_API + "003",
```
Comment (vishwa-vyom) [inline registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/exception/util/PlatformErrorMessages.java]: why is 003 repeated ?

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/token/validation/TokenValidator.java [hunk 3]
```diff
@@ -176,6 +176,11 @@ public boolean validateAccess(String url, String role) {
 				if (role.contains(assignedRole))
 					return true;
 			}
+		} else if (url.contains("userNotification")) {
+			for (String assignedRole : APIAuthorityList.USERNOTIFICATION.getList()) {
+				if (role.contains(assignedRole))
+					return true;
+			}
```
Comment (vishwa-vyom) [inline registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/token/validation/TokenValidator.java:185]: Are we not using properties like the below for authorization checks ?

mosip.role.registration.getsearchrid

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/WorkflowManagerApplication.java [hunk 4]
```diff
@@ -35,5 +36,7 @@ public static void main(String[] args) {
 		workflowActionApi.deployVerticle();
 		WorkflowActionJob workflowActionJob = ctx.getBean(WorkflowActionJob.class);
 		workflowActionJob.deployVerticle();
+		UserNotificationApi messageSenderApi = ctx.getBean(UserNotificationApi.class);
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/WorkflowManagerApplication.java]: can we change the variable name also to match the class name ? This variable looks like we are carrying the old baggage.

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/config/WorkflowManagerConfigBeans.java [hunk 5]
```diff
@@ -89,4 +91,14 @@ public WorkflowActionJob getWorkflowActionJob() {
...
 		return new PublisherClientImpl<WorkflowPausedForAdditionalInfoEventDTO>();
 	}
+	
+	@Bean
+	public UserNotificationApi getUserNotificationApi() {
+		return new UserNotificationApi();
+	}
+	
+	@Bean
+	public UserNotificationRequestValidator getMessageSenderRequestValidator() {
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/config/WorkflowManagerConfigBeans.java]: why is the method name after the message sender ?

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/validator/UserNotificationRequestValidator.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	private static final String RID = "rid";
+
+	/** The Constant ID_FIELD. */
+	private static final String REGTYPE = "regType";
+
+
+	private static final String USER_NOTIFICATION_ID = "mosip.regproc.user.notification.api.id";
+
+
+	private static final String MESSAGE_SENDER_VERSION = "mosip.regproc.user.notification.api.version";
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/validator/UserNotificationRequestValidator.java]: Lets change the name here also

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+	@Value("${mosip.registration.processor.datetime.pattern}")
+	private String dateTimePattern;
+
+	@Value("${mosip.regproc.user.notification.api.id}")
+	private String id;
+
+	@Value("${mosip.regproc.user.notification.api.version}")
+	private String version;
+	
+	private static final String STAGE_PROPERTY_PREFIX = "mosip.regproc.reprocessor.";
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: Why is this property called our as reprocessor ?

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+		return Integer.parseInt(eventBusPort);
+	}
+
+	/**
+	 * contains all the routes in this stage
+	 *
+	 * @param router
+	 */
+	private void routes(MosipRouter router) {
+		router.post(contextPath + "/userNotification");
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: Keep N as small n, as per the convention no camel case in the API path

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+	 */
+	public void processURL(RoutingContext ctx) {
+		LogDescription description = new LogDescription();
+		InternalRegistrationStatusDto registrationStatusDto;
+		JsonObject obj = ctx.getBodyAsJson();
+		String rid = null;
+		try {
+			UserNotificationDTO UserNotificationDTO = JsonUtil.readValueWithUnknownProperties(obj.toString(),
+					UserNotificationDTO.class);
+			UserNotificationRequestDTO UserNotificationRequestDTO = UserNotificationDTO.getRequest();
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: "UserNotificationRequestDTO" and "UserNotificationDTO" - camel case not followed

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+			rid = UserNotificationRequestDTO.getRid();
+			regProcLogger.debug("UserNotificationApi:processURL called for registration id {}",
+					UserNotificationDTO.getRequest().getRid());
+			UserNotificationRequestValidator.validate(UserNotificationDTO);
+			String user = getUser(ctx);
+
+			registrationStatusDto = registrationStatusService.getRegistrationStatus(rid,UserNotificationRequestDTO.getRegType(),1,null);
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: Why is 1 hardcoded here ?

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+			}
+			if (!registrationStatusDto.getRegistrationType().equals(UserNotificationRequestDTO.getRegType())) {
+				description.setMessage(PlatformErrorMessages.RPR_UNA_REGTYPE_NOT_MATCHING.getMessage());
+				description.setCode(PlatformErrorMessages.RPR_UNA_REGTYPE_NOT_MATCHING.getCode());
+				updateAudit(description, UserNotificationRequestDTO.getRid(), false, user);
+				throw new UserNotificationException(PlatformErrorMessages.RPR_UNA_REGTYPE_NOT_MATCHING.getCode(),
+						String.format(PlatformErrorMessages.RPR_UNA_REGTYPE_NOT_MATCHING.getMessage(), rid));
+				}
+				
+				sendWorkflowCompletedWebSubEvent(registrationStatusDto);
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java:183]: Sending workflow completed websub event will not cause other actions additional to resending the notification ?

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+	 * @param registrationId          the registration id
+	 * @param isTransactionSuccessful the is transaction successful
+	 */
+	private void updateAudit(LogDescription description, String registrationId, boolean isTransactionSuccessful,
+			String user) {
+
+		String moduleId = isTransactionSuccessful ? MODULE_ID : description.getCode();
+
+		String eventId = isTransactionSuccessful ? EventId.RPR_402.toString() : EventId.RPR_405.toString();
+		String eventName = isTransactionSuccessful ? EventName.UPDATE.toString() : EventName.EXCEPTION.toString();
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java:241]: In this audit for resent user notification, event name as udpate and related eventId is not making sense

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+	}
+
+	/**
+	 * @param rid
+	 * @param errorCode
+	 * @param errorMessage
+	 * @param e
+	 * @param ctx
+	 */
+	private void logError(String rid, String errorCode, String errorMessage,
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: This method is not only logging error, but also resending back an error response, so name should convey the both

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/resources/bootstrap.properties [hunk 14]
```diff
@@ -1,4 +1,4 @@
-spring.cloud.config.uri=http://localhost:51000/config
+spring.cloud.config.uri=localhost
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/resources/bootstrap.properties:1]: Is this change necessary ?

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/resources/bootstrap.properties [hunk 15]
```diff
@@ -9,9 +9,13 @@ vertx.cluster.configuration=${spring.cloud.config.uri}/*/${spring.profiles.activ
 #management.security.enabled=false
 #disabling health check so that client doesnt try to load properties from sprint config server every
 # 5 minutes (should not be done in production)
+#UserNotification api version
+mosip.regproc.user.notification.api.version=1.0
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/resources/bootstrap.properties]: I don't see the property of workflow action api for the same api version (mosip.regproc.workflow-manager.action.version), do we really nead this here ?

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/test/java/io/mosip/registration/processor/workflowmanager/service/test/UserNotificationApiTest.java [hunk 16]
```diff
@@ -0,0 +1,522 @@
...
+
+            @Override
+            public @Nullable Buffer getBody() {
+                return null;
+            }
+
+//            @Override
+//            public Buffer getBody() {
+//                return null;
+//            }
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/test/java/io/mosip/registration/processor/workflowmanager/service/test/UserNotificationApiTest.java]: If these kind of commented code is not required, please remove the same

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/test/java/io/mosip/registration/processor/workflowmanager/validator/test/UserNotificationRequestValidatorTest.java [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+	}
+
+	@Test(expected = UserNotificationRequestValidationException.class)
+	public void testInValidVersion() throws UserNotificationRequestValidationException {
+		UserNotificationDTO userNotificationDTO = new UserNotificationDTO();
+		userNotificationDTO.setId("mosip.registration.processor.user.notification");
+		userNotificationDTO.setVersion("1.1");
+		userNotificationRequestValidator.validate(userNotificationDTO);
+
```
Comment (vishwa-vyom) [inline registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/test/java/io/mosip/registration/processor/workflowmanager/validator/test/UserNotificationRequestValidatorTest.java]: Lets remove all the unwanted new lines

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/constant/APIAuthorityList.java (hunk 1 above)
Comment (Khuddusshariff0022) [reply registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/constant/APIAuthorityList.java]: Added REGISTRATION_ADMIN.

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 11 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java:183]: No it will not make any other action. only publish the event to websub.

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 12 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java:241]: RPR_402 is Id for update event and RPR_405 is Id for exception event.

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/resources/bootstrap.properties (hunk 14 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/resources/bootstrap.properties:1]: Yes. these are Default properties. we change it to http://localhost:51000/config when we setup in our local system.

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/token/validation/TokenValidator.java (hunk 3 above)
Comment (Khuddusshariff0022) [reply registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/token/validation/TokenValidator.java:185]: No. here we are not using this property.

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/exception/util/PlatformErrorMessages.java (hunk 2 above)
Comment (Khuddusshariff0022) [reply registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/exception/util/PlatformErrorMessages.java]: changed to error code to "005"

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/WorkflowManagerApplication.java (hunk 4 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/WorkflowManagerApplication.java]: Done

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/config/WorkflowManagerConfigBeans.java (hunk 5 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/config/WorkflowManagerConfigBeans.java]: changed method name to "getUserNotificationRequestValidator()"

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/validator/UserNotificationRequestValidator.java (hunk 6 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/validator/UserNotificationRequestValidator.java]: Done

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 8 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: Done

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 9 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: Done

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 13 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: Method name changed to "logErrorAndSendNotifivation"

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 7 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: changed the property to "mosip.regproc.workflow-manager."

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/test/java/io/mosip/registration/processor/workflowmanager/validator/test/UserNotificationRequestValidatorTest.java (hunk 17 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/test/java/io/mosip/registration/processor/workflowmanager/validator/test/UserNotificationRequestValidatorTest.java]: Removed

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/test/java/io/mosip/registration/processor/workflowmanager/service/test/UserNotificationApiTest.java (hunk 16 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/test/java/io/mosip/registration/processor/workflowmanager/service/test/UserNotificationApiTest.java]: removed

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/resources/bootstrap.properties (hunk 15 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/resources/bootstrap.properties]: removed

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/token/validation/TokenValidator.java (hunk 3 above)
Comment (vishwa-vyom) [reply registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/token/validation/TokenValidator.java:185]: Can we create a JIRA tasks to take care of this in the future and the share the id here ?

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 10 above)
Comment (vishwa-vyom) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: I don't find any response to this

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 13 above)
Comment (vishwa-vyom) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: notification - spell check

File: registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/token/validation/TokenValidator.java (hunk 3 above)
Comment (Khuddusshariff0022) [reply registration-processor/registration-processor-core/src/main/java/io/mosip/registration/processor/core/token/validation/TokenValidator.java:185]: https://mosip.atlassian.net/browse/MOSIP-31259

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 13 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: Done

File: registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java (hunk 10 above)
Comment (Khuddusshariff0022) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: it is not being used in dao layer to fetch data. in all other cases it is a part of message DTO. so message DTO is not being used here so kept as 1.

Comment (vishwa-vyom) [reply registration-processor/workflow-engine/registration-processor-workflow-manager-service/src/main/java/io/mosip/registration/processor/workflowmanager/verticle/UserNotificationApi.java]: We have to check the behaviour of iteration and design this accordingly so it make sense in the user notification feature. As per the code it takes the 0th element without order from the list and uses that to send the notification. 
Please discuss with @sowmya695

**Comment by sowmya695** [comment]
Hi @Khuddusshariff0022 ,
this PR please close  as 1.2.0.1 released and raise against develop branch only by giving this as reference for review comment.
Iteration need to discuss with @vishwa-vyom .
cc @vadherhemant
